### PR TITLE
refactor(app): results screen shows calibration and recommendations

### PR DESCRIPTION
**Pull request by sfoster1**
In testing, users couldn't really figure out good next steps from
problems in the cal check recommendations; this way, we not only say
exactly what calibrations we're talking about but also explicitly tell
people to recalibrate rather than just saying a calibration is bad.
Screenshot from 2020-11-12 16-41-46

### CODE REVIEW

**Review by ahiuchingau — APPROVED** [review]
I like the new layout. 

But I personally feel that adding the name of the pipette & the tip rack would make the result summary easier to read

**Comment by Laura-Danielle** [comment]
> I like the new layout.
> 
> But I personally feel that adding the name of the pipette & the tip rack would make the result summary easier to read

Definitely agree with Alise here

**Comment by sfoster1** [comment]
It feels that way, but pretty much every one of the user testers had no clue about next steps or how the errors related to the underlying calibration model. We definitely need to have this information (which specific calibrations are affected) present.

I'm honestly not sure what having the pipette models and tiprack names gives us. @emilywools what do you think?

**Comment by emilywools** [comment]
ill make a mockup and try adding back in the tip rack and pipette names to see if we can make it work with both

**Review by Laura-Danielle — APPROVED** [review]
lgtm

File: app/src/components/CheckCalibration/ResultsSummary.js [hunk 1]
```diff
@@ -103,73 +132,114 @@ export function ResultsSummary(props: CalibrationPanelProps): React.Node {
...
+          />
+        </Box>
       </Flex>
-      <Box paddingX="5%">
-        <Flex marginBottom={SPACING_4}>
-          <Box>
-            <Text marginBottom={SPACING_2}>{DECK_CALIBRATION_HEADER}</Text>
-            <RenderResult status={deckCalibrationResult} />
+      <Flex marginBottom={SPACING_4}>
+        <Box>
```
Comment (b-cooper) [inline app/src/components/CheckCalibration/ResultsSummary.js]: either this `Box` or the `Flex` that wraps it is extraneous

File: app/src/components/CheckCalibration/ResultsSummary.js [hunk 2]
```diff
@@ -103,73 +132,114 @@ export function ResultsSummary(props: CalibrationPanelProps): React.Node {
...
+              textTransform={TEXT_TRANSFORM_UPPERCASE}
+              fontSize={FONT_SIZE_BODY_2}
+              marginBottom={SPACING_3}
+            >
+              {`${m} ${MOUNT}`}
+            </Text>
+            <Box
+              backgroundColor={OVERLAY_LIGHT_GRAY_50}
+              paddingX="5%"
+              height={SPACING_7}
```
Comment (b-cooper) [inline app/src/components/CheckCalibration/ResultsSummary.js]: the `SPACING_X` constants are generally for things like `margin`, `padding`, other layout rules. the `SIZE_X` constants are generally used for sizing `width`, `height`, `minWidth`etc

File: app/src/components/CheckCalibration/ResultsSummary.js [hunk 3]
```diff
@@ -103,73 +132,114 @@ export function ResultsSummary(props: CalibrationPanelProps): React.Node {
...
-          })}
-        </Flex>
-      </Box>
+        ))}
+      </Flex>
       <Box>
-        <Flex
-          alignItems={ALIGN_CENTER}
-          justifyContent={JUSTIFY_CENTER}
+        <Text
```
Comment (b-cooper) [inline app/src/components/CheckCalibration/ResultsSummary.js]: Mike pointed me toward a cleaner pattern that uses the `Btn` primitive here instead it looks like this: 

```
      <Btn color={C_BLUE} onClick={handleDownloadButtonClick} fontSize={FONT_SIZE_BODY_2} marginBottom={SPACING_5}>
        {DOWNLOAD_SUMMARY}
      </Btn>
```

File: app/src/components/CheckCalibration/ResultsSummary.js [hunk 4]
```diff
@@ -235,12 +312,74 @@ function PipetteResult(props: PipetteResultProps): React.Node {
...
+}
+
+function WarningText(props: {|
+  deckCalibrationBad: boolean,
+  pipettes: {|
+    left: {| offsetBad: boolean, tipLengthBad: boolean |},
+    right: {| offsetBad: boolean, tipLengthBad: boolean |},
+  |},
+|}): React.Node {
+  const badCount = reduce(
```
Comment (b-cooper) [inline app/src/components/CheckCalibration/ResultsSummary.js]: The array prototype in JS has it's own reduce function, so you can save the lodash import. e.g.

```
  [
      props.deckCalibrationBad,
      props.pipettes.left.offsetBad,
      props.pipettes.left.tipLengthBad,
      props.pipettes.right.offsetBad,
      props.pipettes.right.tipLengthBad,
    ].reduce(
      (sum, item) => sum + (item  ? 1 : 0),
      0
    )
```

File: app/src/components/CheckCalibration/ResultsSummary.js [hunk 5]
```diff
@@ -235,12 +312,74 @@ function PipetteResult(props: PipetteResultProps): React.Node {
...
+      props.pipettes.right.tipLengthBad,
+    ],
+    (sum, item) => sum + (item === true ? 1 : 0),
+    0
+  )
+  return badCount > 0 ? (
+    <>
+      <Box marginTop={SPACING_3} fontSize={FONT_SIZE_BODY_2}>
+        <Text display={DISPLAY_INLINE_BLOCK}>
+          {`${FOUND}`}
```
Comment (b-cooper) [inline app/src/components/CheckCalibration/ResultsSummary.js:350]: you can avoid the `nbsp`'s, display inline-block's, and extra `Text` tags, by making this one string interpolation e.g.

```
<Text>
    {`${FOUND} ${badCount === 1 ? ONE_ISSUE : PLURAL_ISSUES}. 
    ${RECALIBRATE_AS_INDICATED}. ${NEED_HELP} `}
    <Link
          display={DISPLAY_INLINE_BLOCK}
          color={C_BLUE}
          external={true}
          href={SUPPORT_URL}
     >
          {CONTACT_SUPPORT}
     </Link>
     {` ${FOR_HELP}.`}
</Text>
```

File: app/src/components/CheckCalibration/ResultsSummary.js [hunk 6]
```diff
@@ -235,12 +312,74 @@ function PipetteResult(props: PipetteResultProps): React.Node {
...
+      {props.deckCalibrationBad ? (
+        <Text
+          fontSize={FONT_SIZE_BODY_1}
+          marginTop={SPACING_2}
+          fontStyle={FONT_STYLE_ITALIC}
+        >{`${NOTE}: ${
+          badCount > 1 ? DO_DECK_FIRST : ''
+        } ${DECK_INVALIDATES}`}</Text>
+      ) : (
+        <React.Fragment />
```
Comment (b-cooper) [inline app/src/components/CheckCalibration/ResultsSummary.js]: can these two empty `React.Fragments` be changed to `null`

File: app/src/components/CheckCalibration/__tests__/ResultsSummary.test.js [hunk 7]
```diff
@@ -35,7 +51,37 @@ describe('ResultsSummary', () => {
...
 
-  const getSaveLink = wrapper => wrapper.find('a').at(1)
+  const getSaveLink = wrapper =>
+    wrapper
+      .find('[children="Download detailed JSON Calibration Check summary"]')
+      .find('a')
+  const getDeckParent = wrapper => wrapper.children(Flex).at(1)
+  const getPipParent = wrapper =>
+    wrapper
+      .children(Flex)
```
Comment (b-cooper) [inline app/src/components/CheckCalibration/__tests__/ResultsSummary.test.js]: these kinds of wandering finders in component tests are good fodder for throwing a `title` attribute (prop) on the targeted component and then just doing a `wrapper.find('MyComponent[title="myTitle"]')`. It cleans up this part of the test

**Review by b-cooper — APPROVED** [review]
Additions look good. I have a few comments, mostly small stylistic tweaks

File: app/src/components/CheckCalibration/ResultsSummary.js (hunk 5 above)
Comment (sfoster1) [reply app/src/components/CheckCalibration/ResultsSummary.js:350]: Yeah I had it that way before but figured it was better without the logic inside the string interp, but if you like it I can change it back.

File: app/src/components/CheckCalibration/__tests__/ResultsSummary.test.js (hunk 7 above)
Comment (sfoster1) [reply app/src/components/CheckCalibration/__tests__/ResultsSummary.test.js]: Oh thank god I was really struggling for a way to do this cleanly and didn't know what prop to use

File: app/src/components/CheckCalibration/ResultsSummary.js (hunk 5 above)
Comment (b-cooper) [reply app/src/components/CheckCalibration/ResultsSummary.js:350]: I'm fine with a ternary that switches between two words in a string interp. Especially if it saves a bunch of `nbsp`'s and unnecessary `<p>` tags

File: app/src/components/CheckCalibration/__tests__/ResultsSummary.test.js [hunk 8]
```diff
@@ -189,13 +218,9 @@ describe('ResultsSummary', () => {
 
   it('saves the calibration report when the button is clicked', () => {
     const wrapper = render()
+    console.log(getSaveLink(wrapper).debug())
```
Comment (b-cooper) [inline app/src/components/CheckCalibration/__tests__/ResultsSummary.test.js]: left over log statement?

Comment (sfoster1) [reply app/src/components/CheckCalibration/__tests__/ResultsSummary.test.js]: Yes indeed

**Pull request merged**